### ISSUE TITLE
Make go-to-definition yield all possible definitions for virtual method invocations

### ISSUE THREAD

**Issue by sethfowler** [sethfowler/pygmalion#12]
Right now it just goes to the base class definition, but this usually isn't useful. It should yield a list of all methods that override the base class method.

**Comment by sethfowler** [comment]
We also need it to work the other way for "find callers" and "find references". Right now using "find callers" on a virtual method will often return nothing, because all of the callers are associated with the base class version of the method.

**Comment by sethfowler** [comment]
This won't really work right until pure virtual functions show up in overrides, alas.

**Comment by sethfowler** [comment]
OK, pure virtual functions actually do show up in overrides, they just had no definition. I added code to libclang to detect pure virtual functions, which I need to upstream, and I've now got pure virtual functions in the list. This means this is now implementable.

**Comment by sethfowler** [comment]
Alright, I have a very basic implementation of this. There are some issues:

(1) Need to sort the results.
(2) Definitely need to filter out cases where I know we aren't using dynamic binding. The most obvious case is invocations on instances rather than through pointers.

**Comment by sethfowler** [comment]
Another issue: I'm pretty sure that right now if `A` has two subclasses `B` and `C`, `C` has subclass `D`, and `A::foo()` is overridden by `B::foo()` and `D::foo()` but there is no `C::foo()`, "find callers" on `C::foo()` will wrongly yield callers of `B::foo()` because a call to `C::foo()` will be resolved as going to `A::foo()`. This needs to be fixed.

**Comment by sethfowler** [comment]
For issue (2) above, two notes:
- We also want to filter invocations that explicitly call a particular override, of the form `Base::foo()` or `instance->Base::foo()`.
- I don't think we can do the filter properly without changes to libclang, because the `->` vs `.` distinction isn't exposed (or rather, it's explicitly exposed as an `UnexposedExpr`).

**Comment by sethfowler** [comment]
Actually I think we can pull this off, because we can determine whether the type of the called object is a pointer or not. If it's not a pointer, we should treat the call as non-virtual.

**Comment by sethfowler** [comment]
Argh. The libclang docs strike again. Just noticed `clang_Cursor_isDynamicCall`.

**Comment by sethfowler** [comment]
Unfortunately, that isn't good enough. It nicely takes care of `Base::foo()` and similar, but does not deal with `instance.foo()`, because from the point of view of the C++ spec that's a dynamic call. In general what we want to do here is ensure that the expression directly under the `MemberRefExpr` for the call is a `MemberRefExpr` or `DeclRefExpr` of kind `Record` (and _not_ `Pointer`). Note that `libclang` treats references as standard `Record`s. However, this is complicated by the fact that `(*p).foo()` has this property, and this is precisely equivalent to `p->foo()`. (ignoring operator overloading, at least)

One solution is to be extremely conservative: if the expression directly under the `MemberRefExpr` for the call is a `MemberRefExpr` or a `DeclRefExpr` _and_ has kind `Record`, we treat the call as non-dynamic. This is always safe for our purposes, but it does mean that there are cases we could've treated as non-dynamic that we don't - for example if a function returns a record by reference and we then call a virtual method on that reference in the same expression, this will be treated as a dynamic call. Unfortunately I think that with some C++ coding styles this will happen a lot. A simple extension that should handle a lot of additional cases is to support a `CallExpr` with kind `Record` as the target as well.

I've considered doing this precisely but I'm inclined to avoid that initially because things get complicated fast once all the various types of C++ expressions get involved. Right now I want to try to get most of the benefit as quickly as possible. I'll file a followup issue about taking a more precise approach, but it won't make 1.0.

**Comment by sethfowler** [comment]
Drat, none of this is STILL safe unless I can distinguish between references and normal values. I'll need to extend libclang to make this visible, or else I'll just have to treat everything as a dynamic call.

**Comment by sethfowler** [comment]
Alright, thanks to the addition of `clang_Cursor_getBaseExpression` in my libclang fork I've taken care of the `MemberRefExpr`, `DeclRefExpr`, and `CallExpr` cases, handling references correctly.

The remaining problem that needs to be solved is the "no method implementation on the concrete type" problem, described above using the example of classes A, B, and C. I think the right move here is to store overrides in the database transitively, extract the type of the call target (again using `getBaseExpression`) and store that in the Refs table, and modify the queries to check the overrides table for possible definitions to ensure that their inheritance hierarchy includes the call target class.

**Comment by sethfowler** [comment]
So at this point this is partially working. The big problems are:
- Some things are getting marked as DynamicCallExpr's that shouldn't be. In particular we appear to not be handling stack instance variables correctly right now.
- The "no method implementation on the concrete type" problem continues to not be handled correctly. There's a plan in the comment above but it needs to be put into action.

**Comment by sethfowler** [comment]
Now all that's left is to fix the "no method implementation on the concrete type" problem.

**Comment by sethfowler** [comment]
There's now type information recorded about each reference, but we need more information to solve this: we need a context for definitions similar to that for references. Then a definition is returned if:
- It is an override of the "base" definition.
- Its context is type T, where T is an override of the original reference's "via" type (the type of the message receiver).

The question is, though: how to deal with this efficiently? It's worth just getting it working first and then worrying about performance, but this is a very inefficient query with the current database design. It'd help _a lot_ if the Overrides table was transitive.

**Comment by sethfowler** [comment]
This is now fixed on the `fix-virtual-methods` branch, and will be closed as soon as I merge that.

Regarding the question in the previous comment about performance, though, things have changed since I wrote that, because SQLite now supports recursive queries via recursive CTEs. The implementation needs to be rewritten to use them (and my pull request for `direct-sqlite` needs to be merged) but it's quite likely that that will eliminate any performance issue.
